rumur/doc/rumur.1
.TH RUMUR 1
.SH NAME
rumur \- Yet another explicit state model checker
.SH SYNOPSIS
.B \fBrumur\fR \fBoptions\fR \fB\-\-output\fR \fIFILE\fR [\fIFILE\fR]
.SH DESCRIPTION
Rumur is a reimplementation of the model checker CMurphi with improved
performance and a slightly different feature set.
.SH OPTIONS
\fB\-\-bound\fR \fISTEPS\fR
.RS
Set a limit for state space exploration. The verifier will stop checking beyond
this depth. A bound of \fB0\fR, the default, indicates unlimited exploration.
That is, the verifier will not stop checking until it has expanded all seen
states.
.RE
.PP
\fB\-\-colour\fR [\fBauto\fR | \fBoff\fR | \fBon\fR]
.RS
Enable or disable the use of ANSI colour codes in the verifier's output. The
default is \fBauto\fR, to auto-detect based on whether the verifier's stdout is
a TTY.
.RE
.PP
\fB\-\-counterexample\-trace\fR [\fBdiff\fR | \fBfull\fR | \fBoff\fR]
.RS
Set how counterexample traces are printed when an error is found during
checking. \fBdiff\fR, the default, prints each state showing only the
differences from the previous state. \fBfull\fR shows the entire contents of
each state. \fBoff\fR disables counterexample trace printing altogether.
.RE
.PP
\fB\-\-deadlock\-detection\fR [\fBoff\fR | \fBstuck\fR | \fBstuttering\fR]
.RS
Enable or disable deadlock detection. Rumur has the ability to generate a
verifier that notices when there is no valid transition out of a state and raise
an error in this scenario. The possible modes for this check are:
.RS
.IP \[bu] 2
\fBoff\fR No deadlock checks are performed.
.IP \[bu]
\fBstuck\fR A deadlock is reached when arriving at a state from which there
are no enabled transitions, and an error is signaled in this case.
.IP \[bu]
\fBstuttering\fR A deadlock is reached in the same circumstances as for the
\fBstuck\fR option or additionally if there are enabled transitions but these
all result in an identical state. For CMurphi users, this is the scenario that
CMurphi considers to be a deadlock.
.RE
.PP
This defaults to \fBstuttering\fR. However, whether such a deadlock actually
represents a problem depends on the properties of the system you are modelling.
Hence you may want to change deadlock detection.
.RE
.PP
\fB\-\-debug\fR or \fB\-d\fR
.RS
Enable debugging options in the generated verifier. This includes enabling
runtime assertions. This will also output debugging messages while generating
the verifier.
.RE
.PP
\fB\-\-help\fR
.RS
Display this information.
.RE
.PP
\fB\-\-max\-errors\fR \fICOUNT\fR
.RS
Number of errors the verifier should report before considering them fatal. By
default this is \fI1\fR, that is exit as soon as an error is encountered.
However, you may wish to set a higher value to get multiple error traces from a
single run.
.RE
.PP
\fB\-\-monopolise\fR
.RS
Assume that the machine the generated verifier will run on is the current host
and that it will be the only process running. This flag causes the verifier to
upfront allocate a seen set that will eventually occupy all of memory. That is,
it is the same as passing \fB\-\-set\-expand\-threshold 100\fR and
\fB\-\-set\-capacity\fR with the total amount of physical memory available on
the current machine.
.RE
.PP
\fB\-\-output\fR \fIFILE\fR or \fB\-o\fR \fIFILE\fR
.RS
Set path to write the generated C verifier's code to.
.RE
.PP
\fB\-\-output\-format\fR [\fBmachine\-readable\fR | \fBhuman\-readable\fR]
.RS
Change the format in which the verifier displays its output. By default, it uses
\fBhuman\-readable\fR which results in progress output and then a final summary
of the result. Using \fBmachine\-readable\fR generates output in an XML format
suitable for consumption by a following tool in an I/O pipeline.
.RE
.PP
\fB\-\-pack\-state\fR [\fBon\fR | \fBoff\fR]
.RS
Set whether auxiliary state data is compressed in the generated verifier.
Compression (\fBon\fR, the default) saves memory at the expense of runtime. That
is, by default the verifier will try to minimise memory usage. If your model is
small enough to comfortably fit in available memory, you may want to set this to
\fBoff\fR to accelerate the checking process.
.RE
.PP
\fB\-\-pointer\-bits [\fBauto\fR | \fIBITS\fR]
.RS
Number of relevant (non-zero) bits in a pointer on the target platform on which
the verifier will be compiled. This option can be used to tune pointer
compression, which can save memory when checking larger models. With the
default, \fBauto\fR, the target platform is used to decide this value. On
x86\-64 Linux, 5\-level paging is assumed, meaning pointers can be compressed
and stored in 56 bits. On 64-bit ARM Linux, it is assumed pointers can be stored
in 48 bits. Other platforms currently have no auto-detection, and will store
pointers uncompressed at their full width.
.PP
If you know a certain number of high bits of pointers on your target platform
are always zero, you can teach Rumur this information with this option. For
example, if you are compiling on an x86\-64 platform that you know is using
4\-level paging you can pass \fB\-\-pointer\-bits\fR \fB48\fR to tell Rumur that
the upper 16 bits of a pointer will always be zero.
.RE
.PP
\fB\-\-quiet\fR or \fB\-q\fR
.RS
Don't output any messages while generating the verifier.
.RE
.PP
\fB\-\-reorder\-fields\fR [\fBon\fR | \fBoff\fR]
.RS
Control whether access to state variables and record fields is optimised by
reordering them. By default this is \fBon\fR, causing the order of a model's
state variables and fields within record types to be optimised to more likely
result in naturally aligned memory accesses, which are assumed to be faster. You
should never normally have cause to turn this \fBoff\fR, but this feature was
buggy when first implemented so this option is provided for debugging purposes.
.RE
.PP
\fB\-\-sandbox\fR [\fBon\fR | \fBoff\fR]
.RS
Control whether the generated verifier uses your operating system's sandboxing
facilities to limit its own operations. The verifier does not intentionally
perform any malicious or dangerous operations, but at the end of the day it is a
generated program that you are going to execute. To safeguard against a bug in
the code generator, it is recommended to constrain the operations the verifier is
allowed to perform if you are using a model you did not write yourself. By
default this is \fBoff\fR.
.RE
.PP
\fB\-\-scalarset\-schedules\fR [\fBon\fR | \fBoff\fR]
.RS
Enable or disable tracking of the permutation of scalarset values for more
comprehensible counterexample traces. The permuting of scalarset values that is
performed during symmetry reduction leads to paths in the state space where a
single scalarset identity does not have the same value throughout the trace.
When this option is \fBon\fR (the default), Rumur tracks these permutations and
takes them into account when printing scalarset values or reconstructing
counterexample traces. The result is more intuitive and easily understandable
traces. Turning this \fBoff\fR may gain runtime performance on models that use
scalarsets. However counterexample traces will likely be confusing in this
configuration, as scalarset variables will appear to have their values change
arbitrarily.
.RE
.PP
\fB\-\-set\-capacity\fR \fISIZE\fR or \fB\-s\fR \fISIZE\fR
.RS
The size of the initial set to allocate for storing seen states. This is given
in bytes and is interpreted to mean the desired size of the set when it is
completely full. That is, the initial allocation performed will be for a number
of "state slots" that, when all occupied, will consume this much memory. Default
value for this 8MB.
.RE
.PP
\fB\-\-set\-expand\-threshold\fR \fIPERCENT\fR or \fB\-e\fR \fIPERCENT\fR
.RS
Expand the state set when its occupancy exceeds this percentage. Default is
\fI75\fR, valid values are \fI1\fR - \fI100\fR. Setting a value of 100 will
result in the set only expanding when completely full. This may sound ideal, but
will actually result in a much longer runtime.
.RE
.PP
\fB\-\-symmetry\-reduction\fR [\fBoff\fR | \fBheuristic\fR | \fBexhaustive\fR]
.RS
Enable or disable symmetry reduction. Symmetry reduction is an optimisation that
decreases the state space that must be searched by deriving a canonical
representation of each state. While two states may not be directly equal, if
their canonical representations are the same only one of them need be expanded.
To take advantage of this optimisation you must be using named \fBscalarset\fR
types. The available options are:
.RS
.IP \[bu] 2
\fBoff\fR Do not use symmetry reduction. All scalarsets will be treated as if
they were range types.
.IP \[bu]
\fBheuristic\fR Use a symmetry reduction algorithm based on sorting the state
data. This is not guaranteed to find a single, canonical representation for each
equivalent state, but it is fast and performs reasonably well empirically. Using
this option, you may explore more states than you need to, with the advantage
that you will process each individual state much faster than with
\fBexhaustive\fR. This is the default.
.IP \[bu]
\fBexhaustive\fR Use a symmetry reduction algorithm based on exhaustive
permutation of the state data. This is guaranteed to find a single, canonical
representation for each equivalent state, but is typically very slow. Use this
if you want to minimise memory usage at the expense of runtime.
.RE
.RE
.PP
\fB\-\-threads\fR \fICOUNT\fR or \fB\-t\fR \fICOUNT\fR
.RS
Specify the number of threads the verifier should use. If you do not specify this
parameter or pass \fI0\fR, the number of threads will be chosen based on the
available hardware threads on the platform on which you generate the model.
.RE
.PP
\fB\-\-trace\fR \fICATEGORY\fR
.RS
Enable tracing of specific events while checking. This option is for debugging
Rumur itself, and lets you generate a verifier that writes events to stderr.
Available event categories are:
.RS
.IP \[bu] 2
\fIhandle_reads\fR Reads from variable handles
.IP \[bu]
\fIhandle_writes\fR Writes to variable handles
.IP \[bu]
\fImemory_usage\fR Summary of memory allocation during checking
.IP \[bu]
\fIqueue\fR Events relating to the pending state queue
.IP \[bu]
\fIset\fR Events relating to the seen state set
.IP \[bu]
\fIsymmetry_reduction\fR Events related to the symmetry reduction optimisation
.IP \[bu]
\fIall\fR Enable all of the above
.RE
.PP
More than one of these can be enabled at once by passing the \fB\-\-trace\fR
argument multiple times. Note that enabling tracing will significantly slow the
verifier and is only intended for debugging purposes.
.RE
.PP
\fB\-\-value\-type\fR \fITYPE\fR
.RS
Change the C type used to represent scalar values in the generated verifier.
Valid values are \fIauto\fR and the C fixed-width types, \fIint8_t\fR,
\fIuint8_t\fR, \fIint16_t\fR, \fIuint16_t\fR, \fIint32_t\fR, \fIuint32_t\fR,
\fIint64_t\fR, and \fIuint64_t\fR. The type you select is mapped to its fast
equivalent (e.g. \fIint_fast8_t\fR) and then used in the verifier. The default
is \fIauto\fR that selects the narrowest type that can contain all the scalar
types in use in your model. It is possible that your model does some arithmetic
that temporarily exceeds the bound of any declared type in your model, in which
case you will need to use this option to select a wider type. However, this is
not a common case.
.RE
.PP
\fB\-\-verbose\fR or \fB\-v\fR
.RS
Output informational messages while generating the verifier.
.RE
.PP
\fB\-\-version\fR
.RS
Display version information and exit.
.RE
.SH SMT OPTIONS
If you have a Satisfiability Modulo Theories (SMT) solver installed, Rumur can
use it to optimise your model while generating a verifier. This functionality is
not enabled by default, but you can use the following options to configure Rumur
to find and use your SMT solver. Some examples of solver configuration:
.PP
.RS
# for Z3 with a 5 second timeout
.br
\fBrumur \-\-smt\-path z3 \-\-smt\-arg=\-smt2 \-\-smt\-arg=\-in \-\-smt\-arg=\-t:5000 ...\fR
.PP
# for CVC4 with a 5 second timeout
.br
\fBrumur \-\-smt\-path cvc4 \-\-smt\-prelude "(set\-logic AUFLIA)" \-\-smt\-arg=\-\-lang=smt2 \-\-smt\-arg=\-\-rewrite\-divk \-\-smt\-arg=\-\-tlimit=5000 ...\fR
.RE
.PP
For other solvers, consult their manpages or documentation to determine what
command line parameters they accept. Then use the options described below to
instruct Rumur how to use them. Note that Rumur can only use a single SMT
solver and specifying the \fB\-\-smt\-path\fR option multiple times will only
retain the last path given.
.PP
\fB\-\-smt\-arg\fR \fIARG\fR
.RS
A command line argument to pass to the SMT solver. This option can be given
multiple times and arguments are passed in the order listed. E.g. if you specify
\fB\-\-smt\-arg=\-\-tlimit\fR \fB\-\-smt\-arg=5000\fR the solver will be called
with the command line arguments \fB\-\-tlimit\fR \fB5000\fR.
.RE
.PP
\fB\-\-smt\-bitvectors\fR [\fBoff\fR | \fBon\fR]
.RS
Select whether simple types (enums, ranges, and scalarsets) are translated to
bitvectors or unbounded integers when passed to the solver. By default,
unbounded integers are used (\fB\-\-smt\-bitvectors off\fR). If you turn this
option on, 64\-bit vectors are used instead. Whether integers, bitvectors, or
both are supported will depend on your solver as well as the SMT logic you are
using.
.RE
.PP
\fB\-\-smt\-budget\fR \fIMILLISECONDS\fR
.RS
Total time allotted for running the SMT solver. That is, the time the solver
will be allowed to run for over multiple executions. This defaults to
\fI30000\fR, 30 seconds. So if the solver runs for 10 seconds the first time it
is called, then 5 seconds the second time it is called, then 20 seconds the
third time it is called, it will not be called again. Note that Rumur trusts the
SMT solver to limit itself to a reasonable timeout per run, so its final run can
exceed the budget. You may want to use the \fB\-\-smt\-arg\fR option to pass the
SMT solver a timeout limit if it supports one.
.RE
.PP
\fB\-\-smt\-path\fR \fIPATH\fR
.RS
Command or path to the SMT solver. This will use your environment's \fBPATH\fR
variable, so if the solver is in one of your system directories you can simply
provide the name of its binary. Note that this option has no effect unless you
also pass \fB\-\-smt\-simplification\fR \fBon\fR.
.RE
.PP
\fB\-\-smt\-prelude\fR \fITEXT\fR
.RS
Text to emit when communicating with the solver prior to sending the actual
problem itself. You can use this to set a solver logic or other options. This
option can be given multiple times and each argument will be passed to the
solver on a separate line.
.RE
.PP
\fB\-\-smt\-simplification\fR [\fBoff\fR | \fBon\fR]
.RS
Disable or enable using the SMT solver to simplify the input model. By default,
this is automatic, in that it is turned \fBon\fR if you use any of the other SMT
options or \fBoff\fR if you do not use them.
.RE
.SH AUTHOR
All comments, questions and complaints should be directed to Matthew Fernandez
<matthew.fernandez@gmail.com>.
.SH LICENSE
This is free and unencumbered software released into the public domain.

Anyone is free to copy, modify, publish, use, compile, sell, or
distribute this software, either in source code form or as a compiled
binary, for any purpose, commercial or non-commercial, and by any
means.

In jurisdictions that recognize copyright laws, the author or authors
of this software dedicate any and all copyright interest in the
software to the public domain. We make this dedication for the benefit
of the public at large and to the detriment of our heirs and
successors. We intend this dedication to be an overt act of
relinquishment in perpetuity of all present and future rights to this
software under copyright law.

THE SOFTWARE IS PROVIDED \[lq]AS IS\[rq], WITHOUT WARRANTY OF ANY KIND,
EXPRESS OR IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF
MERCHANTABILITY, FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT.
IN NO EVENT SHALL THE AUTHORS BE LIABLE FOR ANY CLAIM, DAMAGES OR
OTHER LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE,
ARISING FROM, OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR
OTHER DEALINGS IN THE SOFTWARE.

For more information, please refer to <http://unlicense.org>


CHANGELOG.rst
Change log
==========

v2026.09.13
-----------

User-facing changes
~~~~~~~~~~~~~~~~~~~
* Bug fix: compilation of the generated checker when using the x32 ABI has been
  repaired. This bug was introduced in v2026.08.30 (commits
  9bd4c196af6aa3fefeacab0a828ec7abd8ac24b7,
  dfe5f5b98ff04ab496e83ac3281bcd442566f8c1).
* The auxiliary tools ``murphi2c``, ``murphi2murphi``, ``murphi2smv``,
  ``murphi2uclid``, and ``murphi2xml`` now print their error messages including
  context and colouring for ease of reading (commits
  bb728dd1052cb2605b04a78f4a776be2bba864ce,
  4b6bd3d668057601cad0cdb4751bfb33305f126b,
  35f0f7f388b227bf07431dfb1ff599c660b6f9f5,
  c0c3e715a0d6e79da0425b4db3a64e9ea928fbb5,
  d98255a32cf957b0f9b0d8700741fd010cf9f376,
  3f466fe3ccd670cb97f369985c179b6e939f0099).

Internal changes
~~~~~~~~~~~~~~~~
* The type of a numeric literal (``Number::type()``) is now a range that
  contains only its exact value, rather than an infinite range. I.e. ``42``
  would previously be considered to have a type ``-∞..∞`` whereas it now has the
  type ``42..42`` (commit d6d5bc93ee6ec12292cf440972de05eae28c254f).
* Reproducible builds of Rumur should now be possible without patching the build
  system (commit 3d0e63d8d5b5622d5d198e3b70363743e9089179).

v2026.08.30
-----------

User-facing changes
~~~~~~~~~~~~~~~~~~~
* Bug fix: passing a ``--value-type`` containing spaces to ``murphi2c`` no
  longer results in malformed code (commits
  7feb6e0aa0b2b079c2e4f00cc5ceb519d9dac6e6,
  e38ce1b5f02beb7f996fac5ff5e64673ca323f20,
  a1d416ec999af9c24de218942b35fa3cb7f6fbbb).
* Bug fix: ``murphi-format`` more correctly handles ``==>`` followed by
  ``begin`` with multiple newlines in-between (commit
  70173dbe7adb55faa6de5e3fba24522fa3beb4e6).
* Bug fix: monitoring of subprocesses (SMT solvers) was corrected, avoiding a
  previous possible deadlock (commit 67c545f43eb8a71fc59465d8f2d9227a9dd7d631).
* Bug fix: ``murphi-format`` recognises the command line option ``-h`` (commit
  dc03b1031931030fd0cb9b7074a4ad345dee5bb5).
* Bug fix: ``murphi-format`` recognises the command line option ``-t`` (commit
  9e47b916f8f208a4bfe36ffdafc094caa37745d9).
* Bug fix: when passing a very small number to ``--set-capacity``, malformed
  code is no longer generated (commit 80f9ac2acbe3db43bb09fd8fba4e87b4f85fd57b).
* Bug fix: ``murphi-format`` recognises and responds to failures to rewind file
  pointers (commits 52eed93e005726fd39b7ad7327c35acba1afc386,
  79be8baacb001a3180c77f5ef89ffccb309a5641).
* The ``undefined`` keyword is now supported (commits
  1c419e523f625b8ff4d31cd33f9a9fb021f6ed42,
  dd3bc59f781c8ae587f3a5753b80ed06ac033c8b).
* The ``union`` type is now supported during parsing and type checking. It is
  still not possible to generate a checker from a model containing unions but
  some auxiliary tools like ``murphi2xml`` fully support unions. The main
  purpose of this addition is to support third-party code using librumur to
  parse union-containing models (commits
  7ecea100928755fb5f5cb37b638799177b3114c1,
  3b58aa9dbfae7082c01b36368785fea530717e31,
  d6b7949fea8c9da3bdbd08d0311b60fca04683cc,
  62a8c4931edc787401726f2e17d40d4832836d0d,
  c8720b77e298e51350be5486829f16f2c2bf15e8,
  c89bd8afdaa1aad2089cecfb7a77d2e28fab68a4,
  5a2545ee489faf4b730e00797b97979e53b3fab2).
* The ``multiset`` type is now supported during parsing and type checking. It is
  still not possible to generate a checker from a model containing multisets but
  some auxiliary tools like ``murphi2xml`` fully support multisets. The main
  purpose of this addition is to support third-party code using librumur to
  parse multiset-containing models (commits
  9f55ca0ed6fbf0b841130f2c71e5bbf69d4957d0,
  0b3e81a6b425d10f5847e8907cf36487944cc017,
  0639e869deb55f9930e485a5029c82316eafc950,
  1942fbaab08493e2a02a02cb1ef3af40b1901cd9,
  b9d77afccfc01ce9be1287580714c8dd3934493c,
  dbd173eefe7ab88b329330d58c214169133aca46,
  df05e0f8df35146f535cc42cc134cd2efbb47735,
  c12c449c53600e8de09fb40459f67590d8399990,
  ba7239c495f448f60977c6059af2f14d16fb509a,
  8eb2b08949f5c2c97ee3dc5f22594c5ffbd1fc39,
  1076e5f55a0e164824a4e1c23f0ad0fd33528dd1,
  d1cfbdd8a8b1cf96e9f265bcc4061b2ecb26205f,
  d70d11730bd472b36f96fc6539a4a05b2c881fd4,
  67127631f48b6f495d858bae6a52de4b0dcaf0ae,
  e534adfddbfa43ee0d572a96a3aa718d8a7bfec2,
  745ce5eec26b43a123ba27e9f7403fd511022307,
  666b89cdfb62af4d39fb292ebe74625715c7a69b,
  2a686b72ac14948168f58957bff44a7898891cec,
  f6617bef7bab564139822149caafd67b0c722614).
* ``--help`` ignores ``$TMPDIR`` when it points to an unusable directory (commit
  b873e30df7b7f14af3529dd0a2355f8b91d6956e).
* A false positive ``-Wshift-count-overflow`` compiler warning when building the
  generated checker with GCC ≥ 15 has been suppressed (commit
  11347fc61465d5aa7c46fb21721a8cf579739e38).

Internal changes
~~~~~~~~~~~~~~~~
* Bug fix: array indexing (``Element``) is considered impure (``is_pure()``
  returns ``false``) if its indexing expression is impure (commit
  2eb8c356927601e6ac4ef0120746da9ebb702ea8).
* Bug fix: array indexing (``Element``) is considered impure (``is_pure()``
  returns ``false``) if its array expression is impure (commit
  b0aa51f9631457fb9992d0ef19229197678a9c2f).
* Bug fix: record field selection (``Field``) is considered impure
  (``is_pure()`` returns ``false``) if its left-hand side is impure (commit
  d6c04d2a15021338e29e5c80bb940e0993c87a1c).
* The ``lower_bound`` and ``upper_bound`` member functions on AST nodes now
  return ``mpz_class`` values instead of ``std::string`` values (commit
  fd2a59a25af8c4a45f6ec85303c6bad9260a069e).

v2026.03.11
-----------

User-facing changes
~~~~~~~~~~~~~~~~~~~
* Bug fix: when ``&`` or ``|`` were used in combination with expressions whose
  type was a type alias of a range type (as opposed to a range type itself), the
  operator would be inferred to be logical AND/OR. These are now inferred to be
  bitwise AND/OR in this situation, matching the behaviour when type aliases are

rumur/rumur.py
import argparse, logging, sys
from CodeGenerator import to_code
from IRGenerator import to_ir
from Log import log
from Parser import Parser
from OptCF import constant_fold
from OptDCE import eliminate_dead_code
from OptSR import strength_reduce

def run_to_fixed_point(ast, *transformers):
    modified = True
    while modified:
        modified = False
        for t in transformers:
            m, ast = t(ast)
            modified |= m
    return ast

def main():
    parser = argparse.ArgumentParser(
        description='a Murphi model optimiser')
    parser.add_argument('--output', '-o', metavar='FILE',
        type=argparse.FileType('w'), default=sys.stdout,
        help='output file')
    parser.add_argument('input', type=argparse.FileType('r'),
        help='input specification')
    opts = parser.parse_args()

    # XXX
    log.setLevel(logging.DEBUG)

    # Construct the parser.
    try:
        p = Parser()
    except Exception as e:
        print >>sys.stderr, str(e)
        return -1

    # Parse the input.
    try:
        ast = p.parse(opts.input.read())
    except Exception as e:
        print >>sys.stderr, str(e)
        return -1

    # Transform the AST into an intermediate representation.
    global_scope, ir = to_ir(ast)

    # Run optimisations.
    OPT_PIPELINE = [
        strength_reduce,
        constant_fold,
        eliminate_dead_code,
    ]

    for p in OPT_PIPELINE:
        ir = p(ir)

    #ast = run_to_fixed_point(ast, constant_fold)

    # program = lift(ast)

    # try:
    #     env = Environment.Environment()
    #     code = CodeGenerator.generate(env, ast)
    # except Exception as e:
    #     print >>sys.stderr, str(e)
    #     return -1

    # print >>opts.output, code

    #opts.output.write(''.join(to_code(env, ast)))

    code = to_code(global_scope, ir)

    for fragment in code:
        opts.output.write(fragment)

    return 0

if __name__ == '__main__':
    sys.exit(main())
